File: frontend/src/services/briefs.ts
import api from "./api";
import { Brief, BriefCreate, BriefUpdate } from "@/types/brief";

// 409(이미 존재)도 정상 흐름으로 취급하도록 처리
export const createBrief = async (data: BriefCreate): Promise<Brief | null> => {
  try {
    const res = await api.post<Brief>("/briefs", data, {
      // 200/201: 생성 성공, 409: 이미 존재 (예외 던지지 않음)
      validateStatus: (s) => s === 200 || s === 201 || s === 409,
    });
    // 409면 UI에서 안내만 하고 null 반환
    return res.status === 409 ? null : res.data;
  } catch {
    // 네트워크/타임아웃 등은 조용히 null
    return null;
  }
};

export const getBriefByTask = async (taskId: number): Promise<Brief | null> => {
  try {
    const res = await api.get<Brief>(`/briefs/task/${taskId}`, {
      // 200: 존재, 404: 없음 — 둘 다 예외 던지지 않음
      validateStatus: (s) => s === 200 || s === 404,
    });
    return res.status === 200 ? res.data : null; // 404면 null
  } catch {
    // 네트워크/타임아웃 등만 조용히 null
    return null;
  }
};

export const updateBrief = (id: number, data: BriefUpdate) =>
  api.patch<Brief>(`/briefs/${id}`, data).then(r => r.data);

export const deleteBrief = (id: number) =>
  api.delete(`/briefs/${id}`).then(r => r.data);

File: backend/app/routers/briefs.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session, select
from app.db.session import get_session
from app.models import Brief, Task
from app.schemas import BriefCreate, BriefRead, BriefUpdate

from sqlalchemy.exc import IntegrityError

router = APIRouter(prefix="/briefs", tags=["briefs"])

@router.post("", response_model=BriefRead)
def create_brief(payload: BriefCreate, session: Session = Depends(get_session)):
    t = session.get(Task, payload.task_id)
    if not t:
        raise HTTPException(404, "Task not found")
    b = Brief(
        task_id=payload.task_id,
        purpose=payload.purpose,
        success_criteria=payload.success_criteria,
        constraints=payload.constraints,
        priority=payload.priority,
        validation=payload.validation,
    )
    try:
        session.add(b)
        session.commit()
        session.refresh(b)
    except IntegrityError:
        session.rollback()
        # 1:1 제약: 동일 task에 이미 Brief 존재
        raise HTTPException(409, "Brief already exists for this task (1:1)")
    return b

@router.get("", response_model=list[BriefRead])
def list_briefs(session: Session = Depends(get_session)):
    return session.exec(select(Brief)).all()

@router.get("/task/{task_id}", response_model=BriefRead)
def get_brief_by_task(task_id: int, session: Session = Depends(get_session)):
    brief = session.exec(select(Brief).where(Brief.task_id == task_id)).first()
    if not brief:
        raise HTTPException(404, "Brief not found")
    return brief

@router.patch("/{brief_id}", response_model=BriefRead)
def update_brief(brief_id: int, payload: BriefUpdate, session: Session = Depends(get_session)):
    brief = session.get(Brief, brief_id)
    if not brief:
        raise HTTPException(404, "Brief not found")
    if payload.purpose is not None:
        brief.purpose = payload.purpose
    if payload.success_criteria is not None:
        brief.success_criteria = payload.success_criteria
    if payload.constraints is not None:
        brief.constraints = payload.constraints
    if payload.priority is not None:
        brief.priority = payload.priority
    if payload.validation is not None:
        brief.validation = payload.validation
    session.add(brief)
    session.commit()
    session.refresh(brief)
    return brief

@router.delete("/{brief_id}")
def delete_brief(brief_id: int, session: Session = Depends(get_session)):
    brief = session.get(Brief, brief_id)
    if not brief:
        raise HTTPException(404, "Brief not found")
    session.delete(brief); session.commit()
    return {"message": "Brief deleted"}
